api/envoy/api/v2/route/route_components.proto

```diff
@@ -1010,6 +1010,17 @@ message RetryPolicy {
 
   // HTTP headers which must be present in the request for retries to be attempted.
   repeated HeaderMatcher retriable_request_headers = 10;
+
+  // The name of the retry policy. The field is used for pluggable retry policy.
+  string name = 11 [(validate.rules).string = {min_bytes: 1}];
+
+  // Implementation specific configuration which depends on the implementation being instantiated.
+  // See the supported retry policy implementations for further documentation.
+  oneof config_type {
+    google.protobuf.Struct config = 12 [deprecated = true];
+
+    google.protobuf.Any typed_config = 13;
```

Comment:
Does this replace the entire policy above? As such should it really live as a oneof next to the policy definition where that gets defined? Or does this use part of the above policy?

Reply:
The retry policy extension doesn't use above fields in the existing retry policy. I was thinking to use oneof to include existing retry policy and pluggable retry policy, like this:
```protobuf
message RetryPolicy{
// RetryPolicy definition
}
message PluggableRetryPolicy{
// PluggableRetryPolicy definition
}
message RouteAction{
  ...
  oneof retry_policy_speficier{
    RetryPolicy retry_policy = x;
    PluggableRetryPolicy pluggable_retry_policy = y;
  }
}
```

I thought we were going to replace existing retry policy with an extension which means we need to deprecate current retry policy definition in the future. Will using the oneof make the deprecation complicate? If putting them in one message, we could just deprecate fields without deprecating the whole message.

Reply:
I don't think the extension retry policy replaces the entire policy. Some fields in the retry policy are used by other modules, e.g., `per_try_timeout`, `host_selection_retry_max_attempts`.  If we separate the extension from the existing retry policy, we have to provide some default value for existing retry policy.

Reply:
When we separate the extension message from the above message, another problem is the retry plugin. Currently, the extension retry policy interface doesn't include the retry priority and host selection, which means that user cannot integrate the retry policy extension with retry policy extensions. In order to replace the whole message, I think we should make the extension interface the same as the core retry policy interface. 

https://github.com/envoyproxy/envoy/blob/3d162074fca24448334e18fb2fa507ffd7819e7e/include/envoy/router/router.h#L150-L231

At least
```c++
  virtual std::chrono::milliseconds perTryTimeout() const PURE; 
  virtual uint32_t numRetries() const PURE; 
  virtual std::vector<Upstream::RetryHostPredicateSharedPtr> retryHostPredicates() const PURE; 
  virtual Upstream::RetryPrioritySharedPtr retryPriority() const PURE; 
  virtual uint32_t hostSelectionMaxAttempts() const PURE; 
  virtual absl::optional<std::chrono::milliseconds> baseInterval() const PURE; 
  virtual absl::optional<std::chrono::milliseconds> maxInterval() const PURE; 
```

from my perspective. WDYT @mattklein123 @snowp